### PR TITLE
Bug/gsye 117 dispatcher

### PR DESCRIPTION
**Pull request by spyrostz**
Contains various different fixes to correct the behavior of the agents and the storage strategy with regards to future markets. The main fixes are in the following areas:
- Bug fixes in the area dispatcher and agent engine classes, that were the root cause of the main issue (offer / bid delete raised exception during cleanup), because the delete event was not forwarded to all future market areas correctly when the delete offer / bid originated from an area other than Home markets. 
- Refactoring of the StorageStrategy / StorageState classes in order to minimize the number of interactions across the 2, and to not use internal properties of the StorageState e.g. offered_sell_kWh. New methods were created in order to get the available energy for buy and sell, and also for updating the properties of the state on offer / bid , delete and trade. Also made private some methods of the StorageState class (e.g. clamp_energy_to_buy/sell which can be triggered internally in the class and does not need to be exposed)
- One specific future market-related fix on the StorageState, which was to take into account the future offered / traded energy when calculating the available energy for spot and future market slots. 
- Adapted tox configuration in order to be able to run each feature file in its own behave process (contrary to running all feature files in the same behave process as it was done till now). The intent is to decrease the total runtime of the integration tests, since it was observed that on the local machine, the integration tests were executed much faster this way. 

It is more convenient to review each commit separately.

### CODE REVIEW

**Review by Hassan754 — APPROVED** [review]
Looks awesome, Thanks!

File: src/gsy_e/models/strategy/__init__.py [hunk 1]
```diff
@@ -790,7 +790,8 @@ def update_bid_rates(self, market: "TwoSidedMarket", updated_rate: float) -> Non
 
             self.remove_bid_from_pending(market.id, bid.id)
             self.post_bid(market, bid.energy * updated_rate,
-                          bid.energy, attributes=bid.attributes, requirements=bid.requirements,
+                          bid.energy, replace_existing=False, attributes=bid.attributes,
+                          requirements=bid.requirements,
```
[the remark this answers is not in the document]
Comment (spyrostz) [reply src/gsy_e/models/strategy/__init__.py:794]: It is mostly for optimization, and also in order to make this process independent of any other offers / bids that the strategy might have posted (for instance, we may in the future create strategies that post more offers / bids to the ones that the offer / bid updaters provide. In this case, we would not like these orders to be deleted when updating the orders of the TemplateStrategyBid/OfferUpdater)

File: src/gsy_e/models/state.py [hunk 2]
```diff
@@ -625,14 +616,17 @@ def clamp_energy_to_buy_kWh(self, market_slot_time_list):
...
             clamped_energy = max(clamped_energy, 0)
             self.energy_to_buy_dict[time_slot] = clamped_energy
 
     def check_state(self, time_slot):
         """
         Sanity check of the state variables.
         """
+        self._clamp_energy_to_sell_kWh([time_slot])
+        self._clamp_energy_to_buy_kWh([time_slot])
+        self._calculate_soc_for_time_slot(time_slot)
```
[the remark this answers is not in the document]
Comment (spyrostz) [reply src/gsy_e/models/state.py:649]: Indeed the dosctring and method name need to be amended here! Fixing, thanks!

**Review by Hassan754 — APPROVED** [review]

File: src/gsy_e/models/state.py [hunk 3]
```diff
@@ -565,14 +565,19 @@ def set_battery_energy_per_slot(self, slot_length: int) -> None:
...
+            energy + self.pledged_sell_kWh[time_slot] + self.offered_sell_kWh[time_slot]
+            - self.pledged_buy_kWh[time_slot] - self.offered_buy_kWh[time_slot])
+        return energy_balance_kWh - self._battery_energy_per_slot > FLOATING_POINT_TOLERANCE
+
+    def _has_battery_reached_max_charge_power(self, energy: float, time_slot: DateTime) -> bool:
+        """Check whether the storage can withhold the passed energy charge value."""
+        energy_balance_kWh = abs(
+            energy + self.pledged_buy_kWh[time_slot] + self.offered_buy_kWh[time_slot]
+            - self.pledged_sell_kWh[time_slot] - self.offered_sell_kWh[time_slot])
+        return energy_balance_kWh - self._battery_energy_per_slot > FLOATING_POINT_TOLERANCE
```
Comment (hannesdiedrich) [inline src/gsy_e/models/state.py:582]: wow, weird, that this was overlooked for so long. Thanks!

**Review by hannesdiedrich — APPROVED** [review]
Thanks a lot for this. LGTM

**Pull request merged**